### ISSUE TITLE
clean up the messy build script situation

### ISSUE THREAD

**Issue by donbright** [openscad/openscad#377]
per Marius, the build scripts are a bit of a tangle right now. 

try to clean up a little, remove unused stuff, try to consolidate common code a little bit. 

try to make the names more consistent 

add documentation about what the scripts are for, how to use them, etc. 

also, make it easier to build the mingw-cross .... maybe even eliminate the 'setenv' step requirement
##

**Comment by thehans** [comment]
Not sure how much of this original issue is still relevant, I think a lot has changed since it was created. But since its still open, it seems like a good place to discuss plans for build improvements.

I think reducing the number of commands needed from fresh OS install would be worth trying.  Particularly the dependency related scripts.

I'd like to try incorporating the functionality of check-dependencies.sh into cmake, have it spit out a report of missing dependencies if any, and interactively ask how to handle them:  Abort,get missing deps, get all deps, build missing, build all, custom(choose to get/build per dependency, and optionally override default ver# to build)
and cmake would generate a makefile based on the action the user chose.  

I'm curious what other's thoughts on this are

**Comment by kintel** [comment]
I feel that the cmake build system shouldn't be interactive or have other side-effects than creating the necessary files for building OpenSCAD.

Otoh, closing the gap between a dependency checker script and cmake would make sense, to avoid surprises when one is out of sync with the other, or when one isn't fully updated or aware of a certain OS configuration. The new cmake build system should have a significantly cleaner dependency finder, so it might be possible to simply invoke cmake from the `check-dependencies.sh` script as a last step. ..possibly by writing a custom cmake script for this purpose where all the deps stuff is shared with the main cmake script.

**Comment by donbright** [comment]
Agree with Marius, about cmake not having side effects or being interactive.

the dependency scripts need to change and/or go away to meet with the modern world. the linux / bsd  dependency scripts were originally ported by me and @t-paul and others from the ones Marius created for OSX. that was back in a different world. internet access was slower, disk space cost more, computers cost more, there were a handful of dependencies and... also i was building alot on gnu compile farm where i only had user level access. 

also did i mention, we used to worry about the Windows Visual Studio C++ build? yes, there is a bunch of cruft in the build system that nobody has used in several years, that is only there because a handful of people built OpenSCAD on Windows **XP** with MSVCC once upon a time when dinosaurs walked the earth. 

things have changed. there are cheap shell providers, 6-core consumer machines, $200 laptops, so its extremely easy to have root on a machine that can do a build. the number of dependencies has grown extremely large so its no longer maintainable to have these scripts on the linux/bsd side.  it is to the point where you are replicating a significant portion of something like pkg-src, nixbuild, or homebrew (linuxbrew), having to bootstrap up dozens of packages.  and it is pretty much next to impossible to build all these dependencies on Windows with MSVCC, and the scripts that we have now probably wouldnt even work because windows command shell has changed. 

i have tested pkg-src, linuxbrew, nix, and discussed elsewhere on this github. linuxbrew in particular is a good example of why interactive build systems dont work -> it is highly unlikely you will understand the message the system is giving you, even if you do its highly unlikely you know the right answer, and even if you do its highly unlikely the system will be able to use your answer to fix something that is broken.

openscad's **check-dependency** script is no longer maintainable due to increased number of dependencies. it should be deleted.  if you are going to throw out qmake and use cmake... well cmake can do the same thing check-dependencies does, with less code, and more accuracy. another reason to get rid of check-dependencies.

**uni-build-dependencies**, same story. it should be eliminated. unmaintainable. 

unit-get-dependencies can survive, maybe, but it needs to simplify

This AppImage or whatever stuff that @t-paul is doing on OpenSUSE build farm is probably more like the future will look. A pre-prepared well defined set of dependencies with set version numbers that we can rely on. Then, libraries get shipped along with the program. 

Why? We have to think about the interaction complexity of dependenices. If we have two dependencies, and they have 2 versions, then we have four possible combinations of libraries that might bug out against each other. If we have 10 dependencies each of which has 2 versions, that is 1024 possible combinations. there is no way to guarantee behavior, or even test behavior, we are fighting the invisible wall of exponentiation. 

We can't get rid of dependencies either. Water likes going downhill, that is how people like adding dependencies to programs... because people like features. We can't add features without adding dependencies. 

nixbuild seems extremely promising as another way to build, because it can guarantee dependency versions on a "machine within a machine" as well, but has some opengl bugs to work out. i have tested the nixbuild branch on about a dozen different linux versions without any major problem. (to do this, VMs are your friend)

to sum up ---> dont be afraid to delete crap and replace it with something better. dont be afraid to deprecate crap and hide it.

**Comment by kintel** [comment]
Closing as this has stalled and lots of improvements were made over the last few years.
